analysis/druidrestoration/src/modules/core/Mastery.ts

```diff
@@ -0,0 +1,334 @@
+  spellAttributions: MasteryAttributionsBySpell = {};
+
+  // Tracks healing attributable to mastery buffs
+  buffAttributions: MasteryAttributionsByBuff = {};
+
+  constructor(options: Options) {
+    super(options);
+
+    // inits spellAttributions with an entry for each HoT that works with Mastery
+    Object.entries(DRUID_HEAL_INFO).forEach(([guid, buff]: [string, HealerSpellInfo]) => {
+      if (buff.masteryStack) {
+        this.spellAttributions[guid] = new MasterySpellAttribution();
+      }
+    });
+
+    this.addEventListener(Events.heal.by(SELECTED_PLAYER), this.onHeal);
+    this.addEventListener(Events.absorbed.by(SELECTED_PLAYER), this.onAbsorbed);
+
+    // for outputting final computed values when debug is enabled
+    DEBUG && this.addEventListener(Events.fightend, this.onFightEnd);
+  }
+
+  onHeal(event: HealEvent): void {
+    const spellId = event.ability.guid;
+    const target = this.combatants.getEntity(event);
+    const healVal = new HealingValue(event.amount, event.absorbed, event.overheal);
+
+    if (target === null) {
+      return;
+    }
+
+    if (this.spellAttributions[spellId]) {
+      this.spellAttributions[spellId].direct += healVal.effective;
+    }
+
+    if (getSpellInfo(spellId).mastery) {
+      const hotsOn = this.getHotsOn(target);
+      const numHotsOn = hotsOn.length;
+      const decomposedHeal = this._decompHeal(healVal, numHotsOn);
+
+      this.totalNoMasteryHealing += decomposedHeal.noMastery;
+      this.druidSpellNoMasteryHealing += decomposedHeal.noMastery;
+      this.masteryTimesHealing += decomposedHeal.noMastery * decomposedHeal.effectiveStackBenefit;
+
+      // tally benefits for spells
+      hotsOn
+        .filter((hotOn) => hotOn !== spellId) // don't double count
+        .forEach((hotOn) =>
+          this._tallyMasteryBenefit(hotOn.toString(), spellId.toString(), decomposedHeal.oneStack),
+        );
+
+      // tally benefits for ratings buffs
+      this.selectedCombatant
+        .activeBuffs()
+        .filter(
+          (buff) =>
+            this.statTracker.statBuffs[buff.ability.guid] &&
+            this.statTracker.statBuffs[buff.ability.guid].mastery,
+        )
+        .forEach((buff) => {
+          const buffId = buff.ability.guid;
+          const statBuff = this.statTracker.statBuffs[buffId];
+          if (!this.buffAttributions[buffId]) {
+            this.buffAttributions[buffId] = new MasteryBuffAttribution(
+              this.statTracker.getBuffValue(statBuff, statBuff.mastery),
+            );
+          }
+
+          this.buffAttributions[buffId].attributable += calculateEffectiveHealing(
+            event,
+            decomposedHeal.relativeBuffBenefit(this.buffAttributions[buffId].buffAmount),
+          );
+        });
+    } else {
+      this.totalNoMasteryHealing += healVal.effective;
+    }
+  }
+
+  onAbsorbed(event: AbsorbedEvent): void {
+    this.totalNoMasteryHealing += event.amount;
+  }
+
+  onFightEnd(_: FightEndEvent): void {
```

Comment:
Do we need to pass an event argument to this method since we aren't using it and the listener is already listening for the FightEndEvent?

Reply:
I think you're right that JS allows this to be written with no arguments. It was a stylistic choice to use the underscore indicating "I acknowledge I was passed the event, but I won't be using it"

Reply:
I wasn't sure if the linter would complain about it or not. I've seen it done both ways and have no preference, was more of a 'appease the linter overlord' comment.

Reply:
The linter automatically runs on pull requests. The "all checks have passed" on this PR indicates it's passed all linting.